### PR TITLE
ENH: Store fitted attributes as dpt and return dpt tensors

### PR DESCRIPTION
**Pull request by fcharras**
~~The compatibility layer for having the sklearn test keep working is still TODO.~~ Let alone that, I think this would be the final state of `KMeansEngine`.

### CODE REVIEW

**Comment by fcharras** [comment]
I wanted to duplicate the `KMeansEngine` class with `KMeansEngineDebug` class that would be loaded when engine `sklearn_numba_dpex.debug` is used rather than bare `sklearn_numba_dpex`. The difference between `debug` and vanilla would be that `debug` engine converts all fitted attributes and all outputs to numpy arrays. The usecase is using it in `sklearn` unit tests.  

But with current plugin design in https://github.com/scikit-learn/scikit-learn/pull/24497 I don't see a way around. One module can only implement one engine. 

Maybe there's a workaround consisting in exposing several top level namespaces (e.g `sklearn_numba_dpex` and `sklearn_numba_dpex_debug`) in the same package using the `packages` list in `setup.py` ? 

Or WDYT about changing the plugin spec to enable a different syntax that allow passing dotted string to the `config_context` ?

E.g rewriting https://github.com/scikit-learn/scikit-learn/pull/24497/files#diff-1d31de81e903bd6529fbe68f8009b7113e3b7de4f1465572ef88af4d03a7dc5bR35 such that entry points whose value prefixes match the user inputed string are selected. (i.e. using prefix selection rather than `str.split(".", 1)` to detect the name of the engine)

**Comment by ogrisel** [comment]
I think  we should allow to programmatically register additional provider names (without in addition the `importlib.metadata` registry) for testing / debug / benchmarking purposes.

**Comment by fcharras** [comment]
Out of WIP

**Review by jjerphan — COMMENTED** [review]
Thanks @fcharras. Here are a few comments.

File: sklearn_numba_dpex/kmeans/engine.py [hunk 1]
```diff
@@ -197,22 +198,28 @@ def kmeans_single(self, X, sample_weight, centers_init_t):
...
+        # depends on device support for float64), shouldn't we cast to a dtype that
+        # is always consistent with the input ? (e.g. cast to float64 if the input
+        # was given as float64 ?) But what assumptions can we make on the input
+        # so we can infer its input dtype without risking triggering a copy of it ?
+        return assignments_idx, inertia, best_centroids_t.T, n_iteration
+
+    def is_same_clustering(self, labels, best_labels, n_clusters):
+        if self._testing_mode:
+            return super().is_same_clustering(labels, best_labels, n_clusters)
+        return is_same_clustering(labels, best_labels, n_clusters)
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:217]: Maybe ordering methods definitions with all engine method put first under some inline comment, followed by private ?

File: sklearn_numba_dpex/kmeans/kernels/utils.py [hunk 2]
```diff
@@ -249,3 +250,84 @@ def reduce_centroid_data(
...
+            return
+
+        label = labels[sample_idx]
+
+        if clusters_seen[label] > zero_idx:
+            return
+
+        is_new = dpex.atomic.add(clusters_seen, zero_idx, one_incr)
+        if is_new == one_incr:
+            dpex.atomic.add(nb_distinct_clusters, zero_idx, one_incr)
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/utils.py:330]: TY for pointing out the bug, i'll add a test. (it wasn't a blocking bug, only responsible for spamming warnings)

For the first question, not sure what you mean. This code ensure that each unique value will be counted once. The documentation reads:

> Returns the old value at the index location **as** if it is loaded atomically.

maybe you read too fast and miss *as* ? :smile:

File: sklearn_numba_dpex/kmeans/kernels/utils.py [hunk 3]
```diff
@@ -249,3 +250,84 @@ def reduce_centroid_data(
...
+        labels2,
+        mapping,
+        result
+    ):
+        # fmt: on
+        """`result` is expected to be an empty array with dtype int32 of size (1,)
+        initialized with value 1.
+        """
+        sample_idx = dpex.get_global_id(zero_idx)
+        if sample_idx >= n_samples:
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/utils.py:299]: Seems better but note that this add a read operation and doesn't stop future tasks, it justs ensure that future tasks will exit early after the read. So on average it's better when clusterings are not the same but not when clusterings are the same. Still better since different clusterings should be the most common outcome.

File: sklearn_numba_dpex/kmeans/kernels/utils.py [hunk 4]
```diff
@@ -249,3 +250,84 @@ def reduce_centroid_data(
                 empty_clusters_list[current_n_empty_clusters] = cluster_idx
 
     return reduce_centroid_data[global_size, work_group_size]
+
+
+@lru_cache
+def make_is_same_clustering_kernel(n_samples, n_clusters, work_group_size, device):
+    # TODO: are there possible optimizations for this kernel ?
+    # - fusing the two kernels (requires a lock ?)
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/utils.py:258]: Can add a comment.

There's a risk of concurrency issues between threads that try to write different values into `mapping` and threads that read `mapping`, all this at the same time.

I've wrapped my mind around this problem trying to fuse it using atomic add/sub but I don't think it's possible. You're welcome to try :-D

File: sklearn_numba_dpex/kmeans/engine.py [hunk 5]
```diff
@@ -89,7 +89,16 @@ def __init__(self, estimator):
                 "https://github.com/IntelPython/numba-dpex/issues/767"
             )
         self.order = order
-        super().__init__(estimator)
+        self.estimator = estimator
+
+        _testing_mode = os.getenv("SKLEARN_NUMBA_DPEX_TESTING_MODE", "0")
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:94]: `True/False` maybe ? I'd prefer not hint that other values could be supported.

File: sklearn_numba_dpex/kmeans/engine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:217]: Or maybe having all methods that do not override the engine methods be prefixed by underscore ? (I think it's good already ?)

File: sklearn_numba_dpex/kmeans/kernels/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/kernels/utils.py:330]: Yes, `is_previous` is a bad name and this must be reverted.

Still, I think that it is possible that two threads read a `0` value and increment the variable twice in independent atomic operations.

[the remark this answers is not in the document]
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/kernels/utils.py:330]: Yes. :upside_down_face:

File: sklearn_numba_dpex/kmeans/engine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/engine.py:217]: Yes. We could have public methods as front-end, some extra operations for the testing mode and then a call to private methods as back-end which would be unchanged.

What do you think?

File: sklearn_numba_dpex/kmeans/engine.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/engine.py:94]: That's fair. In this case should we opt for the alternative I've proposed, keeping `0/1` as values for `SKLEARN_NUMBA_DPEX_TESTING_MODE`?

File: sklearn_numba_dpex/kmeans/engine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/engine.py:217]: Well yes but that's already how the class works, just needs to add a comment that document it.

**Comment by fcharras** [comment]
Comments should be addressed in last commit and suggestions applied.

File: .github/workflows/run_tests.yaml [hunk 6]
```diff
@@ -27,5 +27,14 @@ jobs:
       run: pytest sklearn_numba_dpex/
 
       # TODO: run those tests in a separate pipeline
+      # NB: `sklearn_numba_dpex` engine set the estimators to output arrays of type
+      # `dpctl.tensor.usm_ndarray` and store fitted attributes with this same type.
+      # This behavior is not compatible with sklearn unit tests, that expects numpy
+      # arrays or at least arrays that would closely mimic the numpy api. To keep
```
Comment (jjerphan) [inline .github/workflows/run_tests.yaml:33]:
```suggestion
      # arrays or at least arrays that would closely mimic the NumPy Python API. To keep
```

File: sklearn_numba_dpex/kmeans/engine.py [hunk 7]
```diff
@@ -61,6 +61,12 @@ class KMeansEngine(KMeansCythonEngine):
 
     """
 
+    # NB: by convention, all methods that are not prefixed with an underscore `_` in
+    # this class (let's call it *public methods*) override the corresponding method
+    # from sklearn.cluster._kmeans.KMeansCythonEngine. All methods that do not override
+    # a method in sklearn.cluster._kmeans.KMeansCythonEngine are considered private to
+    # this implementation and prefixed with `_`.
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/engine.py:68]: Thank you for adding this comment!

As private and public methods are conventionally defined as such, I think this can be boiled down to:

```suggestion
    # NB: by convention, all public methods override the corresponding method
    # from `sklearn.cluster._kmeans.KMeansCythonEngine`.
    # All methods that do not override a method in
    # `sklearn.cluster._kmeans.KMeansCythonEngine` are considered private to
    # this implementation.
```

What do you think?

File: sklearn_numba_dpex/kmeans/engine.py [hunk 8]
```diff
@@ -91,14 +97,14 @@ def __init__(self, estimator):
...
             raise ValueError(
-                "If the environment variable SKLEARN_NUMBA_DPEX_TESTING_MODE is set, "
-                f'it is expected to take values in {"0", "1"}, but got {_testing_mode}'
-                " instead."
+                "If the environment variable SKLEARN_NUMBA_DPEX_is_testing_mode is"
+                f' set, it is expected to take values in {"0", "1"}, but got'
+                f" {_is_testing_mode} instead."
             )
-        self._testing_mode = _testing_mode == "1"
+        self._is_testing_mode = _is_testing_mode == "1"
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/engine.py:107]: Ah, I meant the following which I think reads better.
```suggestion
        self._is_in_testing_mode = _is_testing_mode == "1"
```

WDYT?

File: sklearn_numba_dpex/kmeans/tests/test_kmeans.py [hunk 9]
```diff
@@ -490,3 +491,15 @@ def test_error_raised_on_invalid_group_sizes():
...
+def test_get_nb_distinct_clusters_kernel():
+    labels = [0, 1, 0, 2, 2, 7, 6, 5, 3, 3]  # NB: all values up to 7 except 4
+    expected_nb_distinct_clusters = 6
+
+    n_clusters = max(labels)
+    labels = dpt.asarray(labels, dtype=np.int32)
+
+    actual_nb_distinct_clusters = int(get_nb_distinct_clusters(labels, n_clusters))
+
+    assert actual_nb_distinct_clusters == expected_nb_distinct_clusters
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/tests/test_kmeans.py:505]: Thanks!

I guess we could test on randomly generated data and by composing `len` with `np.unique`, but we needn't: this is sufficient IMO.

**Review by jjerphan — APPROVED** [review]
LTGM. Thank you, @fcharras!

**Comment by fcharras** [comment]
So in fact `make_get_nb_distinct_clusters_kernel` does suffer from a race condition, the unittest sometimes output `6`, sometimes `7`, but it more seems like a bug in `numba_dpex` atomics rather than our code.

**Comment by fcharras** [comment]
The test was the issue, we have `n_clusters = max(labels) + 1` but used `n_clusters = max(labels)`.
